modules/xml/xml.c

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (c) 2017 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "xml.h"
+#include "scratch-buffers.h"
+
+typedef struct
+{
+  LogMessage *msg;
+  GString *key;
+  gboolean pop_next_time;
+  XMLParser *parser;
+} InserterState;
+
+static void
+remove_trailing_dot(gchar *str)
+{
+  g_assert(strlen(str));
+  if (str[strlen(str)-1] == '.')
+    str[strlen(str)-1] = 0;
+}
+
+static void
+msg_add_attributes(LogMessage *msg, GString *key, const gchar **names, const gchar **values)
+{
+  GString *attr_key;
+
+  if (names[0])
+    {
+      attr_key = scratch_buffers_alloc();
+      g_string_printf(attr_key, "%s._", key->str);
+    }
+  else
+    return;
+
+  gint base_index = attr_key->len;
+  gint attrs = 0;
+
+  while (names[attrs])
+    {
+      attr_key = g_string_overwrite(attr_key, base_index, names[attrs]);
+      log_msg_set_value_by_name(msg, attr_key->str, values[attrs], -1);
+      attrs++;
+    }
+}
+
+static gboolean
+tag_matches_patterns(const gchar *element_name, const GPtrArray *patterns, gboolean should_reverse)
```

Comment:
What if we pass the `reversed_element_name` also? In this case we don't have to propagate this flag to here (and this decision already made by the caller...). On the other hand in this case it is better to pass also the length of the `element_name` (as the caller also has to compute it because of the reverse).

Reply:
Very good comment thanks! Especially again, element_name is _NOT_ null terminated, so I should not have called strlen() on it inside this function. I will fix this.